json output propery for @robguttman solver

joshbduncan opened the pull request:
@robguttman, here's a start for the json output. You can access it using the `.json` property on a WordSearch object. it includes the puzzle, words, and key. Sample output below. Have a look and let me know what you think? Thanks!

```json
{
	"puzzle": [
		["A", "H", "F", "A", "N", "T", "G", "A", "Y", "M", "O", "Q"],
		["O", "N", "S", "Z", "O", "A", "H", "S", "U", "W", "D", "K"],
		["K", "G", "Y", "K", "X", "N", "D", "Y", "Z", "C", "R", "M"],
		["I", "Z", "U", "R", "D", "L", "I", "B", "K", "L", "D", "B"],
		["M", "Q", "C", "T", "N", "J", "S", "G", "U", "Z", "X", "E"],
		["C", "R", "N", "U", "M", "D", "E", "M", "T", "H", "B", "H"],
		["A", "A", "I", "R", "O", "J", "L", "P", "D", "E", "Z", "E"],
		["T", "B", "P", "T", "G", "C", "T", "N", "A", "M", "O", "T"],
		["H", "B", "I", "L", "J", "D", "L", "O", "S", "I", "Y", "X"],
		["X", "I", "G", "E", "A", "W", "V", "P", "W", "J", "U", "E"],
		["M", "T", "D", "O", "N", "K", "E", "Y", "I", "E", "V", "Z"],
		["J", "I", "S", "G", "B", "U", "V", "R", "A", "U", "R", "C"]
	],
	"key": {
		"RABBIT": {
			"start": [5, 1],
			"direction": "S"
		},
		"DONKEY": {
			"start": [10, 2],
			"direction": "E"
		},
		"TURTLE": {
			"start": [4, 3],
			"direction": "S"
		},
		"PIG": {
			"start": [7, 2],
			"direction": "S"
		},
		"CAT": {
			"start": [5, 0],
			"direction": "S"
		}
	}
}
```

robguttman commented:
This looks fantastic, @joshbduncan !  I hope to have time to give it a spin this coming weekend.  Thanks!

robguttman reviewed (approved):
Looks simple and great!

robguttman commented:
@joshbduncan , I tested out the JSON format above and it works wonderfully except for one issue.  The answer key's `start` and `end` values are off by 1.  Although Python's lists are 0-indexed, your textual answers are 1-indexed.  So for example, this answer key:
```
		"RABBIT": {
			"start": [5, 1],
			"direction": "S"
		},
```
actually maps to answer "S @ (6, 2)".  So two suggested changes:

1. have the answer key in JSON match the 1-index answer key shown in text
2. split the `start` key into `start` and `end` since having it as a list isn't as useful and structured as it could be - for example the above answer key example would be better as:
```
		"RABBIT": {
			"start": 6,
			"end": 2,
			"direction": "S"
		},
```
What do you think?

joshbduncan commented:
@robguttman thanks for the suggestions! I just pushed the following changes to this PR...

1. Updated the JSON key to be 1-based like the "non-dev" user-facing output.
2. Added an "end" key to the JSON for each word. The "start" and "end" must be tuples since words can go at an angle.

Have a look and let me know what you think?

robguttman commented:
@joshbduncan , thanks, this is better.  I realize now that I made a mistake in my second suggestion.  It should have been:
```
		"RABBIT": {
			"start-row": 6,
			"start-col": 2,
			"direction": "S"
		},
```
The above is more structured and clearer than using lists each coordinate - and makes extracting the data easier.  It's fine adding `end` keys but they're not strictly needed as they're redundant with the `direction` and `start`.  I won't be using them in the solver but others may find `end` keys useful so no harm keeping it.

joshbduncan commented:
@robguttman, okay that makes way more sense. I didn't quite understand your "start/end" suggestion but it was late so I just took a guess as to what you meant... And yes, since the direction is included, the end coordinates are not needed.

I just pushed an updated version that includes the requested `start_row`, and `start_col`. I'll have to sort out the failing test but all should be good for you to test it. Have a look and let me know if this works better for your solver? Thanks!

robguttman commented:
@joshbduncan , this change looks great!  It works well in the solver.  Good luck with the failing test!

joshbduncan commented:
@robguttman, it was just flask8 complaining about an unused import. All is good now. Thanks again!

joshbduncan commented:
@robguttman, I made some substantial improvements that make sure no duplicate words could end up in the puzzle. There was a very small chance it could happen (especially with very similar words), so now the generator checks to make sure this can't happen every time it inserts a character (either from a word insertion or from a random fill character insertion). This also checks to make sure no palindromes are present either as that would cause confusion.

robguttman commented:
Wow, good thinking and nice work, @joshbduncan !
